### PR TITLE
Support hypre_ParCSRMatrixLocalTranspose

### PR DESCRIPTION
**Pull request by artv3**
Improved communication performance for nc-conforming meshes. 
See: https://github.com/CEED/benchmarks/pull/18

Should be merged after Hypre merges https://github.com/hypre-space/hypre/tree/parcsr_local_trans 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3052](https://github.com/mfem/mfem/pull/3052) | @artv3 | @tzanio | @tzanio + @YohannDudouit + @v-dobrev | 7/5/22 | 7/30/22 | 8/2/22 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

File: linalg/hypre_parcsr.cpp [hunk 1]
```diff
@@ -1697,7 +1697,8 @@ void hypre_ParCSRMatrixBooleanMatvecT(hypre_ParCSRMatrix *A,
 
    if (num_cols_offd)
    {
-#if MFEM_HYPRE_VERSION >= 21100
+#if MFEM_HYPRE_VERSION >= 21100 && 0
```
Comment (liruipeng) [inline linalg/hypre_parcsr.cpp:1700]: Not sure about the `&& 0`. I don't know the code much.

**Review by liruipeng — APPROVED** [review]
Looks good to me as far as I can tell.

File: linalg/hypre_parcsr.cpp (hunk 1 above)
Comment (artv3) [reply linalg/hypre_parcsr.cpp:1700]: That is on us, we are turning it off since there is some memory management stuff to sort out.

**Comment by tzanio** [comment]
GIven that https://github.com/hypre-space/hypre/pull/646 is merged, are we ready to move forward with this?

**Comment by artv3** [comment]
> GIven that [hypre-space/hypre#646](https://github.com/hypre-space/hypre/pull/646) is merged, are we ready to move forward with this?

@liruipeng is there a way to check if we are using Hypre beyond a particular commit? For this PR, we would need the current version from develop (which is a bit ahead of the last release).

**Comment by liruipeng** [comment]
> > GIven that [hypre-space/hypre#646](https://github.com/hypre-space/hypre/pull/646) is merged, are we ready to move forward with this?
> 
> @tzanio yes! The only issue is I think we may need to check Hypre if the function we need is there.
> 
> @liruipeng is there a way to check if we are using Hypre beyond a particular commit? For this PR, we would need the current version from develop (which is a bit ahead of the last release).

How about `HYPRE_DEVELOP_STRING`?

**Comment by artv3** [comment]
> > > GIven that [hypre-space/hypre#646](https://github.com/hypre-space/hypre/pull/646) is merged, are we ready to move forward with this?
> > 
> > 
> > @tzanio yes! The only issue is I think we may need to check Hypre if the function we need is there.
> > @liruipeng is there a way to check if we are using Hypre beyond a particular commit? For this PR, we would need the current version from develop (which is a bit ahead of the last release).
> 
> How about `HYPRE_DEVELOP_STRING`?

If i build the master branch of Hypre I see:  #define HYPRE_DEVELOP_STRING "v2.25.0-1-gb58585e0f"

How can we check if we are using that version or grater? I imagine the hash at the end i just random right? or does the -1- correspond to 1 commit beyond v2.25.0?

**Comment by tzanio** [comment]
You want something like

```
#if (MFEM_HYPRE_VERSION <= 22300) && !(MFEM_HYPRE_VERSION == 22300 && HYPRE_DEVELOP_NUMBER >=8)
```
see e.g. https://github.com/mfem/mfem/blob/master/linalg/hypre.cpp#L1680

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: linalg/hypre_parcsr.cpp [hunk 2]
```diff
@@ -1560,7 +1560,8 @@ void hypre_ParCSRMatrixAbsMatvecT(hypre_ParCSRMatrix *A,
...
```
Comment (tzanio) [inline linalg/hypre_parcsr.cpp:1563]: @artv3, can we take out the `&& 0`?

Comment (artv3) [reply linalg/hypre_parcsr.cpp:1563]: @v-dobrev suggested I commented it out. What do you think @v-dobrev ?

File: linalg/hypre_parcsr.cpp [hunk 3]
```diff
@@ -1560,7 +1560,8 @@ void hypre_ParCSRMatrixAbsMatvecT(hypre_ParCSRMatrix *A,
...
+      //function below will crash if offdT is on the device
```
Comment (tzanio) [inline linalg/hypre_parcsr.cpp:1564]: Can we change this to:

```suggestion
      // Disabled for now, until we implement hypre_CSRMatrixAbsMatvec on device
```

File: linalg/hypre_parcsr.cpp [hunk 4]
```diff
@@ -1575,7 +1576,8 @@ void hypre_ParCSRMatrixAbsMatvecT(hypre_ParCSRMatrix *A,
 
    comm_handle = hypre_ParCSRCommHandleCreate(2, comm_pkg, y_tmp, y_buf);
 
-#if MFEM_HYPRE_VERSION >= 21100
+#if MFEM_HYPRE_VERSION >= 21100 && 0
+   //function below will crash if offdT is on the device
```
Comment (tzanio) [inline linalg/hypre_parcsr.cpp:1580]:
```suggestion
   // Disabled for now, until we implement hypre_CSRMatrixAbsMatvec on device
```

File: linalg/hypre_parcsr.cpp [hunk 5]
```diff
@@ -1697,7 +1699,8 @@ void hypre_ParCSRMatrixBooleanMatvecT(hypre_ParCSRMatrix *A,
 
    if (num_cols_offd)
    {
-#if MFEM_HYPRE_VERSION >= 21100
+#if MFEM_HYPRE_VERSION >= 21100 && 0
+      //function below will crash if offdT is on the device
```
Comment (tzanio) [inline linalg/hypre_parcsr.cpp:1703]:
```suggestion
      // Disabled for now, until we implement hypre_CSRMatrixAbsMatvec on device
```

File: linalg/hypre_parcsr.cpp [hunk 6]
```diff
@@ -1712,7 +1715,8 @@ void hypre_ParCSRMatrixBooleanMatvecT(hypre_ParCSRMatrix *A,
 
    comm_handle = hypre_ParCSRCommHandleCreate_bool(2, comm_pkg, y_tmp, y_buf);
 
-#if MFEM_HYPRE_VERSION >= 21100
+#if MFEM_HYPRE_VERSION >= 21100 && 0
+   //function below will crash if offdT is on the device
```
Comment (tzanio) [inline linalg/hypre_parcsr.cpp:1719]:
```suggestion
   // Disabled for now, until we implement hypre_CSRMatrixAbsMatvec on device
```

**Review by tzanio — APPROVED** [review]

**Comment by v-dobrev** [comment]
The MacOS checks for 351a5a00 "failed" because of this:
```
This is a scheduled macOS-10.15 brownout. The macOS-10.15 environment is deprecated and will
be removed on August 30th, 2022. For more details, see
   https://github.com/actions/virtual-environments/issues/5583
```
We'll need to update the MacOS version we use.

**Comment by v-dobrev** [comment]
@artv3, @tzanio, I made some changes -- see the commit message 9592ee5a. Let me know if those look good to you.

**Review by v-dobrev — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by tzanio — APPROVED** [review]
Thanks Arturo!

**Pull request merged**